cookbooks/aws-parallelcluster-config/files/default/custom_action_executor/custom_action_executor.py

```diff
@@ -0,0 +1,518 @@
+        if self._is_https_url(script.url):
+            return await self._download_http_script(exe_script)
+
+        raise DownloadRunError(
+            f"Failed to download {self.event_name} script {step_num} {script.url}, URL must be an s3 or HTTPs.",
+            f"Failed to download {self.event_name} script {step_num}, URL must be an s3 or HTTPs.",
+        )
+
+    @staticmethod
+    def _build_exe_script(script, step_num, path):
+        return ExecutableScript(script.url, script.args, step_num, path)
+
+    async def _download_s3_script(self, exe_script: ExecutableScript):
+        s3_client = boto3.resource("s3")
+        bucket_name, key = self._parse_s3_url(exe_script.url)
+        with tempfile.NamedTemporaryFile(delete=False) as file:
+            try:
+                s3_client.Bucket(bucket_name).download_file(key, file.name)
+            except (NoCredentialsError, botocore.exceptions.ClientError) as err:
+                os.unlink(file.name)
+                raise DownloadRunError(
+                    f"Failed to download {self.event_name} script {exe_script.step_num} {exe_script.url}"
+                    f" using aws s3, cause: {err}.",
+                    f"Failed to download {self.event_name} script {exe_script.step_num} using aws s3.",
+                ) from err
+            exe_script.path = file.name
+            return exe_script
+
+    async def _download_http_script(self, exe_script: ExecutableScript):
+        url = exe_script.url
+        response = requests.get(url, timeout=DOWNLOAD_SCRIPT_HTTP_TIMEOUT_SECONDS)
+        if response.status_code != 200:
+            raise DownloadRunError(
+                f"Failed to download {self.event_name} script {exe_script.step_num} {url}, "
+                f"HTTP status code {response.status_code}.",
+                f"Failed to download {self.event_name} script {exe_script.step_num} via HTTP.",
+            )
+        with tempfile.NamedTemporaryFile(delete=False) as file:
+            file.write(response.content)
+        exe_script.path = file.name
+        return exe_script
+
+    async def _execute_script(self, exe_script: ExecutableScript):
+        # preserving error case for making the script executable
+        try:
+            subprocess.run(["chmod", "+x", exe_script.path], check=True)  # nosec - trusted input
+        except subprocess.CalledProcessError as err:
+            raise DownloadRunError(
+                f"Failed to run {self.event_name} script {exe_script.step_num} {exe_script.url} "
+                f"due to a failure in making the file executable, return code: {err.returncode}.",
+                f"Failed to run {self.event_name} script {exe_script.step_num} "
+                f"due to a failure in making the file executable, return code: {err.returncode}.",
+            ) from err
+
+        # execute script with it's args
+        try:
+            # arguments are provided by the user who has the privilege to create/update the cluster
```

Comment:
security relevant exclusion: surfacing an issue already present in the bash implementation, see this comment in the code